File: _stats_build/statistics/wr_single_history.rb
require_relative "abstract/round_metric"

# NOTE: WR Single 历史——RoundMetric 的退化情况
# compute_metric 直接返回 best 字段，无需从 value1-5 计算
class WrSingleHistory < RoundMetric
  # NOTE: metric = best 字段本身，用高效两步 SQL 排名（不参与 compute_all_rankings）
  def self.batch_ranking? = false
  def initialize
    @title = "Single"
    @title_zh = "单次"
    @note = "Shows how world record singles have progressed over time for each event."
    @note_zh = "展示各项目世界纪录单次成绩随时间的变化。"
    @table_header = { "Result" => :right, "Improvement" => :right, "Days" => :right,
                      "Person" => :left, "Competition" => :left, "Date" => :left, "Details" => :left }
  end

  # NOTE: Single 的 metric 就是 best 字段本身
  def compute_metric(_values, r) = r["best"]
  def format_metric(v, eid) = SolveTime.new(eid, :single, v.round).clock_format
  def target_events = Events::OFFICIAL
  def wr_record_column = "regional_single_record"
  def value_column = "best"
  def value_type = :single
end

File: trainer/packages/stats-build/src/statistics/wr_single_history.ts
// NOTE: WR Single History——单次 WR 进展历史
// 与 Ruby _stats_build/statistics/wr_single_history.rb 1:1 对应
// compute_metric 直接返回 best 字段，无需从 attempts 计算
import { RoundMetric } from '../core/round_metric.js';
import { OFFICIAL_EVENTS_RECORD } from '../core/events.js';
import { SolveTime } from '../core/solve_time.js';
import type { RowDataPacket } from 'mysql2';

export class WrSingleHistory extends RoundMetric {
  constructor() {
    super();
    this.title = 'Single';
    this.titleZh = '单次';
    this.note = 'Shows how world record singles have progressed over time for each event.';
    this.noteZh = '展示各项目世界纪录单次成绩的历史变化。';
    this.tableHeader = {
      'Result': 'right', 'Improvement': 'right', 'Days': 'right',
      'Person': 'left', 'Date': 'left', 'Competition': 'left', 'Details': 'left',
    };
    // NOTE: 覆写钩子——使用 best 字段和 single 记录列
    this.wrRecordColumn = 'regional_single_record';
    this.valueColumn = 'best';
    this.valueType = 'single';
  }

  // NOTE: batch_ranking = false——用高效两步 SQL
  batchRanking() { return false; }

  targetEvents() { return OFFICIAL_EVENTS_RECORD; }

  computeMetric(_values: number[], row: RowDataPacket): number | null {
    return Number(row['best']);
  }

  formatMetric(v: number, eid: string): string {
    return new SolveTime(eid, 'single', Math.round(v)).clockFormat();
  }
}
